bdai_ros2_wrappers/bdai_ros2_wrappers/executors.py

```diff
@@ -0,0 +1,414 @@
+        logger: typing.Optional[logging.Logger] = None,
+    ):
+        """
+        Initializes the thread pool.
+
+        :param min_workers: minimum number of workers in the pool, 0 by default.
+        :param max_workers: maximum number of workers in the pool, 32 times the
+          number of available CPU threads by default (assuming I/O bound work).
+        :param submission_quota: maximum number of concurrent submissions for a
+          a given callable. Up to the maximum number of workers by default. Useful
+          when serving multiple users to prevent anyone from starving the rest.
+        :param submission_patience: time to wait in seconds for a worker to become
+          available before upscaling the pool. 100 ms by default.
+        :param max_idle_time: time in seconds for a worker to wait for work before
+          shutting itself down, effectively downscaling the pool. 60 seconds by default.
+        """
+
+        if min_workers is None:
+            min_workers = 0
+        if min_workers < 0:
+            raise ValueError("Minimum number of workers must be a nonnegative number")
+        self._min_workers = min_workers
+
+        if max_workers is None:
+            max_workers = 32 * (os.cpu_count() or 1)
+        if max_workers < min_workers:
+            raise ValueError("Maximum number of workers must be larger than or equal to the minimum number of workers")
+        self._max_workers = max_workers
+
+        if max_idle_time is None:
+            max_idle_time = 60.0
+        if max_idle_time <= 0:
+            raise ValueError("Maximum idle time for workers must be a positive number")
+        self._max_idle_time = max_idle_time
+
+        if submission_quota is None:
+            submission_quota = max_workers
+        if submission_quota <= 0:
+            raise ValueError("Quota for submission must be a positive number")
+        self._submission_quota = submission_quota
+
+        if submission_patience is None:
+            submission_patience = 0.1
+        if submission_patience < 0:
+            raise ValueError("Patience for submission must be a nonnegative number")
+        self._submission_patience = submission_patience
+
+        if logger is None:
+            logger = logging.getLogger(classname(self))
+        self._logger = logger
+
+        self._shutdown = False
+        self._submit_lock = threading.Lock()
+        self._shutdown_lock = threading.Lock()
+
+        with AutoScalingThreadPool._lock:
+            if AutoScalingThreadPool._interpreter_shutdown:
+                raise RuntimeError("cannot create thread pool while interpreter is shutting down")
+            # register runqueue for external wake up  on interpreter shutdown
+            runqueue: queue.SimpleQueue = queue.SimpleQueue()
+            AutoScalingThreadPool._all_runqueues.add(runqueue)
+            self._runqueue = runqueue
+            weak_self = weakref.ref(self, lambda ref: runqueue.put(None))
```

Comment:
> is there any chance that any of these weak references get eaten by the garbage collection before the job is done running

It could happen, yes, but that's a non-issue. Once the worker is done working, if the executor is gone, it will self-terminate.

> What happens to the weak_self in the thread after the _do_work function runs?

Things changed a bit since you reviewed, but not fundamentally. We keep a weak reference to the executor to avoid the reference from persisting the pool in a daemonized worker once it can no longer be accessed from anywhere else. We dereference it when necessary and delete the reference explicitly for the same reasons.